### PR TITLE
SOLR-13971: CVE-2019-17558: Velocity custom template RCE vulnerability

### PR DESCRIPTION
**Pull request by Sachpat**
# Description

This is the fix for CVE-2019-17558: Velocity custom template RCE vulnerability, which was fixed for 8.3 and I provided the same fix for 7.7 as per the discussion here:
https://issues.apache.org/jira/browse/SOLR-13971?focusedCommentId=17010489&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17010489

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [ ] I have developed this patch against the `master` branch.
- [x] I have run `ant precommit` and the appropriate test suite.
- [x] I have added tests for my changes.
- [x] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

**Comment by Sachpat** [comment]
@chatman please take a look :)

**Comment by Sachpat** [comment]
@erikhatcher I also committed the changes for SOLR-14025 in this.

File: solr/contrib/velocity/src/java/org/apache/solr/response/VelocityResponseWriter.java [hunk 1]
```diff
@@ -231,25 +222,32 @@ private VelocityContext createContext(SolrQueryRequest request, SolrQueryRespons
...
+      /*
+          // Custom tools, specified in config as:
+              <queryResponseWriter name="velocityWithCustomTools" class="solr.VelocityResponseWriter">
+                <lst name="tools">
+                  <str name="mytool">com.example.solr.velocity.MyTool</str>
+                </lst>
+              </queryResponseWriter>
+      */
+      // Custom tools can override any of the built-in tools provided above, by registering one with the same name
+      if (request.getCore().getCoreDescriptor().isConfigSetTrusted()) {
```
Comment (artem-smotrakov) [inline solr/contrib/velocity/src/java/org/apache/solr/response/VelocityResponseWriter.java]: Looks like it checks `request.getCore().getCoreDescriptor().isConfigSetTrusted()` twice. I suppose one check may be removed.

Comment (Sachpat) [reply solr/contrib/velocity/src/java/org/apache/solr/response/VelocityResponseWriter.java]: Yes, that's right @artem-smotrakov . Actually I tried to align the code according to the original patch https://github.com/apache/lucene-solr/commit/9dfee35b9facbcbc9342d8bbbe52667cb6d1dced.  It was also the same there :) @erikhatcher , should I change it or keep it as it is?

**Comment by chatman** [comment]
Merged. Thanks. Lets close this PR.

**Comment by Sachpat** [comment]
Thanks for merging it @chatman . Will close this PR now.

**Comment by rhtham** [comment]
@chatman I am trying to figure out if the following is a mitigation step for CVE-2019-17558 on SOLR 6.1.  None of our solrconfig.xml contains the lib references to the velocity jar files as follows:

  <lib dir="${solr.install.dir:../../../..}/contrib/velocity/lib" regex=".*\.jar" />
  <lib dir="${solr.install.dir:../../../..}/dist/" regex="solr-velocity-\d.*\.jar" />
  
It doesn't appear that you can add these jars references using the config API.  Without these references, you are not able to flip the params.resource.loader.enabled to true using the config API.  If you are not able to flip the flag and none of your cores have these lib references then is the risk present?

Thanks in advance!

**Comment by erikhatcher** [comment]
No.  Without those libs, no risk present.

**Pull request closed without merging**